pymc_marketing/prior.py

```diff
@@ -0,0 +1,893 @@
+
+
+class MuAlreadyExistsError(Exception):
+    """Error for when 'mu' is present in Prior."""
+
+    def __init__(self, distribution: Prior) -> None:
+        self.distribution = distribution
+        self.message = f"The mu parameter is already defined in {distribution}"
+        super().__init__(self.message)
+
+
+class UnknownTransformError(Exception):
+    """Error for when an unknown transform is used."""
+
+
+def _remove_leading_xs(args: list[str | int]) -> list[str | int]:
+    """Remove leading 'x' from the args."""
+    while args and args[0] == "x":
+        args.pop(0)
+
+    return args
+
+
+def handle_dims(x: pt.TensorLike, dims: Dims, desired_dims: Dims) -> pt.TensorVariable:
+    """Take a tensor of dims `dims` and align it to `desired_dims`.
+
+    Doesn't check for validity of the dims
+
+    Examples
+    --------
+    1D to 2D with new dim
+
+    .. code-block:: python
+
+        x = np.array([1, 2, 3])
+        dims = "channel"
+
+        desired_dims = ("channel", "group")
+
+        handle_dims(x, dims, desired_dims)
+
+    """
+    x = pt.as_tensor_variable(x)
+
+    if np.ndim(x) == 0:
+        return x
+
+    dims = dims if isinstance(dims, tuple) else (dims,)
+    desired_dims = desired_dims if isinstance(desired_dims, tuple) else (desired_dims,)
+
+    aligned_dims = np.array(dims)[:, None] == np.array(desired_dims)
+
+    missing_dims = aligned_dims.sum(axis=0) == 0
+    new_idx = aligned_dims.argmax(axis=0)
+
+    args = [
+        "x" if missing else idx
+        for (idx, missing) in zip(new_idx, missing_dims, strict=False)
+    ]
+    args = _remove_leading_xs(args)
+    return x.dimshuffle(*args)
+
+
+DimHandler = Callable[[pt.TensorLike, Dims], pt.TensorLike]
+
+
+def create_dim_handler(desired_dims: Dims) -> DimHandler:
+    """Wrapper to act like the previous `create_dim_handler` function."""
+
+    def func(x: pt.TensorLike, dims: Dims) -> pt.TensorVariable:
+        return handle_dims(x, dims, desired_dims)
+
+    return func
+
+
+def _dims_to_str(obj: tuple[str, ...]) -> str:
+    if len(obj) == 1:
+        return f'"{obj[0]}"'
+
+    return (
+        "(" + ", ".join(f'"{i}"' if isinstance(i, str) else str(i) for i in obj) + ")"
+    )
+
+
+def _get_pymc_distribution(name: str) -> type[pm.Distribution]:
+    if not hasattr(pm, name):
+        raise UnsupportedDistributionError(
+            f"PyMC doesn't have a distribution of name {name!r}"
+        )
+
+    return getattr(pm, name)
+
+
+def _get_transform(name: str):
+    for module in (pt, pm.math):
+        if hasattr(module, name):
+            break
+    else:
+        module = None
+
+    if not module:
+        raise UnknownTransformError(
+            f"Neither PyTensor or pm.math have the function {name!r}"
+        )
+
+    return getattr(module, name)
+
+
+def _get_pymc_parameters(distribution: pm.Distribution) -> set[str]:
+    return set(signature(distribution.dist).parameters.keys()) - {"kwargs", "args"}
+
+
+class Prior:
```

Comment:
Here is an opinionated suggestion: Inherint from PyDAntic BAse model and use model validators https://docs.pydantic.dev/latest/concepts/validators/#dataclass-validators to do the checks. See also https://docs.pydantic.dev/latest/concepts/validation_decorator/

Given its popularity, I do not think Pydantic is a strong dependency.  WDYT?

Reply:
Think it is a great idea. Let's save off for another PR